File: src/backend/nfe/src/model/nfe_declaracao_importacao_model.py
from src import db


class NfeDeclaracaoImportacaoModel(db.Model):
    __tablename__ = 'nfe_declaracao_importacao'

    id = db.Column(db.Integer, primary_key=True)
    numero_documento = db.Column(db.String(12))
    data_registro = db.Column(db.DateTime)
    local_desembaraco = db.Column(db.String(60))
    uf_desembaraco = db.Column(db.String(2))
    data_desembaraco = db.Column(db.DateTime)
    via_transporte = db.Column(db.String(1))
    valor_afrmm = db.Column(db.Float)
    forma_intermediacao = db.Column(db.String(1))
    cnpj = db.Column(db.String(14))
    uf_terceiro = db.Column(db.String(2))
    codigo_exportador = db.Column(db.String(60))
    id_nfe_detalhe = db.Column(db.Integer, db.ForeignKey('nfe_detalhe.id'))


    def mapping(self, data):
        self.id = data.get('id')
        self.id_nfe_detalhe = data.get('idNfeDetalhe')
        self.numero_documento = data.get('numeroDocumento')
        self.data_registro = data.get('dataRegistro')
        self.local_desembaraco = data.get('localDesembaraco')
        self.uf_desembaraco = data.get('ufDesembaraco')
        self.data_desembaraco = data.get('dataDesembaraco')
        self.via_transporte = data.get('viaTransporte')
        self.valor_afrmm = data.get('valorAfrmm')
        self.forma_intermediacao = data.get('formaIntermediacao')
        self.cnpj = data.get('cnpj')
        self.uf_terceiro = data.get('ufTerceiro')
        self.codigo_exportador = data.get('codigoExportador')

    def serialize(self):
        return {
            'id': self.id,
            'idNfeDetalhe': self.id_nfe_detalhe,
            'numeroDocumento': self.numero_documento,
            'dataRegistro': self.data_registro.isoformat(),
            'localDesembaraco': self.local_desembaraco,
            'ufDesembaraco': self.uf_desembaraco,
            'dataDesembaraco': self.data_desembaraco.isoformat(),
            'viaTransporte': self.via_transporte,
            'valorAfrmm': self.valor_afrmm,
            'formaIntermediacao': self.forma_intermediacao,
            'cnpj': self.cnpj,
            'ufTerceiro': self.uf_terceiro,
            'codigoExportador': self.codigo_exportador,
        }

File: frontend/nfe/lib/app/data/model/nfe_declaracao_importacao_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfe/app/infra/infra_imports.dart';
import 'package:intl/intl.dart';
import 'package:nfe/app/data/domain/domain_imports.dart';

class NfeDeclaracaoImportacaoModel {
	int? id;
	int? idNfeDetalhe;
	String? numeroDocumento;
	DateTime? dataRegistro;
	String? localDesembaraco;
	String? ufDesembaraco;
	DateTime? dataDesembaraco;
	String? viaTransporte;
	double? valorAfrmm;
	String? formaIntermediacao;
	String? cnpj;
	String? ufTerceiro;
	String? codigoExportador;

	NfeDeclaracaoImportacaoModel({
		this.id,
		this.idNfeDetalhe,
		this.numeroDocumento,
		this.dataRegistro,
		this.localDesembaraco,
		this.ufDesembaraco,
		this.dataDesembaraco,
		this.viaTransporte,
		this.valorAfrmm,
		this.formaIntermediacao,
		this.cnpj,
		this.ufTerceiro,
		this.codigoExportador,
	});

	static List<String> dbColumns = <String>[
		'id',
		'numero_documento',
		'data_registro',
		'local_desembaraco',
		'uf_desembaraco',
		'data_desembaraco',
		'via_transporte',
		'valor_afrmm',
		'forma_intermediacao',
		'cnpj',
		'uf_terceiro',
		'codigo_exportador',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Numero Documento',
		'Data Registro',
		'Local Desembaraco',
		'Uf Desembaraco',
		'Data Desembaraco',
		'Via Transporte',
		'Valor Afrmm',
		'Forma Intermediacao',
		'Cnpj',
		'Uf Terceiro',
		'Codigo Exportador',
	];

	NfeDeclaracaoImportacaoModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNfeDetalhe = jsonData['idNfeDetalhe'];
		numeroDocumento = jsonData['numeroDocumento'];
		dataRegistro = jsonData['dataRegistro'] != null ? DateTime.tryParse(jsonData['dataRegistro']) : null;
		localDesembaraco = jsonData['localDesembaraco'];
		ufDesembaraco = NfeDeclaracaoImportacaoDomain.getUfDesembaraco(jsonData['ufDesembaraco']);
		dataDesembaraco = jsonData['dataDesembaraco'] != null ? DateTime.tryParse(jsonData['dataDesembaraco']) : null;
		viaTransporte = NfeDeclaracaoImportacaoDomain.getViaTransporte(jsonData['viaTransporte']);
		valorAfrmm = jsonData['valorAfrmm']?.toDouble();
		formaIntermediacao = NfeDeclaracaoImportacaoDomain.getFormaIntermediacao(jsonData['formaIntermediacao']);
		cnpj = jsonData['cnpj'];
		ufTerceiro = NfeDeclaracaoImportacaoDomain.getUfTerceiro(jsonData['ufTerceiro']);
		codigoExportador = jsonData['codigoExportador'];
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNfeDetalhe'] = idNfeDetalhe != 0 ? idNfeDetalhe : null;
		jsonData['numeroDocumento'] = numeroDocumento;
		jsonData['dataRegistro'] = dataRegistro != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataRegistro!) : null;
		jsonData['localDesembaraco'] = localDesembaraco;
		jsonData['ufDesembaraco'] = NfeDeclaracaoImportacaoDomain.setUfDesembaraco(ufDesembaraco);
		jsonData['dataDesembaraco'] = dataDesembaraco != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataDesembaraco!) : null;
		jsonData['viaTransporte'] = NfeDeclaracaoImportacaoDomain.setViaTransporte(viaTransporte);
		jsonData['valorAfrmm'] = valorAfrmm;
		jsonData['formaIntermediacao'] = NfeDeclaracaoImportacaoDomain.setFormaIntermediacao(formaIntermediacao);
		jsonData['cnpj'] = Util.removeMask(cnpj);
		jsonData['ufTerceiro'] = NfeDeclaracaoImportacaoDomain.setUfTerceiro(ufTerceiro);
		jsonData['codigoExportador'] = codigoExportador;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNfeDetalhe = plutoRow.cells['idNfeDetalhe']?.value;
		numeroDocumento = plutoRow.cells['numeroDocumento']?.value;
		dataRegistro = Util.stringToDate(plutoRow.cells['dataRegistro']?.value);
		localDesembaraco = plutoRow.cells['localDesembaraco']?.value;
		ufDesembaraco = plutoRow.cells['ufDesembaraco']?.value != '' ? plutoRow.cells['ufDesembaraco']?.value : 'AC';
		dataDesembaraco = Util.stringToDate(plutoRow.cells['dataDesembaraco']?.value);
		viaTransporte = plutoRow.cells['viaTransporte']?.value != '' ? plutoRow.cells['viaTransporte']?.value : 'AAA';
		valorAfrmm = plutoRow.cells['valorAfrmm']?.value?.toDouble();
		formaIntermediacao = plutoRow.cells['formaIntermediacao']?.value != '' ? plutoRow.cells['formaIntermediacao']?.value : 'AAA';
		cnpj = plutoRow.cells['cnpj']?.value;
		ufTerceiro = plutoRow.cells['ufTerceiro']?.value != '' ? plutoRow.cells['ufTerceiro']?.value : 'AC';
		codigoExportador = plutoRow.cells['codigoExportador']?.value;
	}	

	NfeDeclaracaoImportacaoModel clone() {
		return NfeDeclaracaoImportacaoModel(
			id: id,
			idNfeDetalhe: idNfeDetalhe,
			numeroDocumento: numeroDocumento,
			dataRegistro: dataRegistro,
			localDesembaraco: localDesembaraco,
			ufDesembaraco: ufDesembaraco,
			dataDesembaraco: dataDesembaraco,
			viaTransporte: viaTransporte,
			valorAfrmm: valorAfrmm,
			formaIntermediacao: formaIntermediacao,
			cnpj: cnpj,
			ufTerceiro: ufTerceiro,
			codigoExportador: codigoExportador,
		);			
	}

	
}